### PR TITLE
Limit maximum engine startup concurrency of kyuubi server

### PR DESCRIPTION
**Pull request by wForget**
### _Why are the changes needed?_

As reported by #4825, a large number of engine builder processes may cause high machine load on the kyuubi server, So I want to add a config to limit engine creation concurrency.

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala [hunk 1]
```diff
@@ -63,6 +63,8 @@ class KyuubiSessionManager private (name: String) extends SessionManager(name) {
   private var batchLimiter: Option[SessionLimiter] = None
   lazy val (signingPrivateKey, signingPublicKey) = SignUtils.generateKeyPair()
 
+  var engineBuilderSemaphore: Option[Semaphore] = None
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:66]:
```suggestion
  var engineStartupProcessSemaphore: Option[Semaphore] = None
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala [hunk 2]
```diff
@@ -72,6 +74,7 @@ class KyuubiSessionManager private (name: String) extends SessionManager(name) {
     addService(credentialsManager)
     metadataManager.foreach(addService)
     initSessionLimiter(conf)
+    initEngineBuilderSemaphore(conf)
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:77]:
```suggestion
    initEngineStartupProcessSemaphore(conf)
```

**Review by pan3793 — APPROVED** [review]

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala [hunk 3]
```diff
@@ -202,7 +203,14 @@ private[kyuubi] class EngineRef(
     }
 
     MetricsSystem.tracing(_.incCount(ENGINE_TOTAL))
+    var acquiredPermit = false
     try {
+      if (!startupProcessSemaphore.forall(_.tryAcquire(timeout, TimeUnit.MILLISECONDS))) {
+        throw KyuubiSQLException(
```
Comment (cxzl25) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala:209]:
```suggestion
        MetricsSystem.tracing(_.incCount(MetricRegistry.name(ENGINE_TIMEOUT, appUser)))
        throw KyuubiSQLException(
```

**Review by cxzl25 — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merged to master

**Comment by risyomei** [comment]
Appreciate your effort for this!

**Pull request closed without merging**